### PR TITLE
[FEATURE] Move operator to spectra merger

### PR DESCRIPTION
**Pull request by lennykovac**
# Description
The SpectraMerger.h class has no member so a move-operator made no sense. Furthermore most of the runtime got spend in _OpenMS::MSSpectrum::sortByPosition_
I did some minor optimazations here and there. 

Please include a summary of the change and which issue is fixed.
Here the Issue: [5621](https://github.com/OpenMS/OpenMS/issues/5621)

# Checklist:
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file -> No relevant changes, just cosmetics.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- `/reformat` (experimental) applies the clang-format style changes as additional commit
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds

### CODE REVIEW

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 1]
```diff
@@ -629,12 +631,12 @@ namespace OpenMS
         }
```
Comment (vrossow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h:639]: In 619: You deleted the same vector further above (228), would it make sense to do the same thing here?

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 2]
```diff
@@ -733,19 +737,19 @@ namespace OpenMS
           if (((*it_mz) - last_mz) > delta_mz)
           {
             mz_positions.push_back(*it_mz);
-            intensities.push_back(0.0);
             last_mz = *it_mz;
           }
         }
+        intensities.resize(mz_positions.size());
```
Comment (hbeschorner) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: Does it really have a performance benefit, I think it is more confusing

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 3]
```diff
@@ -225,16 +225,16 @@ namespace OpenMS
           // make cluster element
           BaseFeature bf;
           bf.setRT(exp[i].getRT());
-          std::vector<Precursor> pcs = exp[i].getPrecursors();
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h:228]: why not make this `const auto& pcs = ...` and leave the rest as is?

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 4]
```diff
@@ -361,7 +360,8 @@ namespace OpenMS
               double weight = 1;
               if (average_type == "gaussian")
               {
-                weight = std::exp(factor * pow(it_rt_2->getRT() - it_rt->getRT(), 2));
+                //factor * (rt_2 -rt)^2
+                weight = std::exp(factor * (it_rt_2->getRT() - it_rt->getRT()) * (it_rt_2->getRT() - it_rt->getRT()));
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]:
```suggestion
                // factor * (rt_2 -rt)^2
               double base = it_rt_2->getRT() - it_rt->getRT();
               weight = std::exp(factor * base * base);
```

Comment (cbielow) [reply src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: just for better readability

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 5]
```diff
@@ -370,7 +370,8 @@ namespace OpenMS
             if (average_type == "gaussian")
             {
               // Gaussian
-              terminate_now = std::exp(factor * pow(it_rt_2->getRT() - it_rt->getRT(), 2)) < cutoff;
+              terminate_now = std::exp(factor * (it_rt_2->getRT() - it_rt->getRT()) *
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: same here

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 6]
```diff
@@ -431,12 +433,12 @@ namespace OpenMS
       for (AverageBlocks::iterator it = spectra_to_average_over.begin(); it != spectra_to_average_over.end(); ++it)
       {
         double sum(0.0);
-        for (std::vector<std::pair<Size, double> >::const_iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2)
+        for (auto it2 = it->second.cbegin(); it2 != it->second.cend(); ++it2)
         {
           sum += it2->second;
         }
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]:
```suggestion
        for (const auto& element : second)
        {
          sum += element.second;
        }
```

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h (hunk 1 above)
Comment (cbielow) [reply src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h:639]:
```
          auto & pcs = consensus_spec.getPrecursors();
          //if (pcs.size()>1) OPENMS_LOG_WARN << "Removing excessive precursors - leaving only one per MS2 spectrum.\n";
          pcs.resize(1);
          pcs[0].setMZ(precursor_mz_average);
   //       consensus_spec.setPrecursors(pcs);
```

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 7]
```diff
@@ -629,12 +631,12 @@ namespace OpenMS
...
         else
         {
-          merged_spectra.addSpectrum(consensus_spec);
+          merged_spectra.addSpectrum(std::move(consensus_spec));
         }
       }
 
       OPENMS_LOG_INFO << "Cluster sizes:\n";
-      for (std::map<Size, Size>::const_iterator it = cluster_sizes.begin(); it != cluster_sizes.end(); ++it)
+      for (auto it = cluster_sizes.cbegin(); it != cluster_sizes.cend(); ++it)
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]:
```suggestion
      for (const auto& cl_size : cluster_sizes)
```

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 8]
```diff
@@ -651,13 +653,15 @@ namespace OpenMS
       {
         if (merged_indices.count(i) == 0) // save unclustered ones
         {
-          exp_tmp.addSpectrum(exp[i]);
+          exp_tmp.addSpectrum(std::move(exp[i]));
           exp[i] = empty_spec;
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: delete

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 9]
```diff
@@ -651,13 +653,15 @@ namespace OpenMS
...
         }
       }
 
       //typedef std::vector<typename MapType::SpectrumType> Base;
       //exp.Base::operator=(exp_tmp);
+      //TODO CLEAN THIS UP!
+      //Meta_Data will not be cleared
       exp.clear(false);
       exp.getSpectra().insert(exp.end(), exp_tmp.begin(), exp_tmp.end());
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: todo: how to move all spectra, instead of copying

Comment (cbielow) [reply src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: https://stackoverflow.com/questions/10720122/is-there-a-standard-way-of-moving-a-range-into-a-vector

Comment (cbielow) [reply src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: same in line `exp.getSpectra().insert(exp.end(), merged_spectra.begin(), merged_spectra.end());`

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 10]
```diff
@@ -765,11 +769,11 @@ namespace OpenMS
           typename MapType::PeakType peak;
           peak.setMZ(mz_positions[i]);
           peak.setIntensity(intensities[i]);
-          average_spec.push_back(peak);
+          average_spec.push_back(std::move(peak));
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: undo, moving is copying

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 11]
```diff
@@ -883,34 +888,26 @@ namespace OpenMS
           typename MapType::PeakType peak;
           peak.setMZ(mz_new[i]);
           peak.setIntensity(intensity_new[i]);
-          average_spec.push_back(peak);
+          average_spec.push_back(std::move(peak));
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: undo

File: src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h [hunk 12]
```diff
@@ -883,34 +888,26 @@ namespace OpenMS
...
       }
 
       logger.endProgress();
 
       // loop over blocks
       int n(0);
-      for (AverageBlocks::const_iterator it = spectra_to_average_over.begin(); it != spectra_to_average_over.end(); ++it)
+      for (auto it = spectra_to_average_over.cbegin(); it != spectra_to_average_over.cend(); ++it)
       {
         exp[it->first] = exp_tmp[n];
```
Comment (cbielow) [inline src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h]: move?

**Review by cbielow — COMMENTED** [review]
with all those changes: does the runtime improve?

**Pull request closed without merging**